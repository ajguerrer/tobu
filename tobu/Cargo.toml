[package]
name = "tobu"
version = "0.1.0"
authors = ["Andrew Guerrero <ajguerrer@gmail.com>"]
edition = "2018"

[dependencies]
bytes = "1"
serde = "1"
thiserror = "1"

tobu-format = { path = "../tobu-format", version = "0.1.0" }


[dev-dependencies]
serde_derive = "1"
