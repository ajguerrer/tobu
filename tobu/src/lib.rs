use bytes::{Bytes, BytesMut};
use error::Error;
use info::MessageInfo;
use ser::{Serializer, SizeHint};
use serde::Serialize;

pub mod de;
pub mod error;
pub mod info;
pub mod ser;
pub mod value;

pub fn to_vec(value: &impl Serialize, info: &'static MessageInfo) -> Result<Vec<u8>, Error> {
    let mut vec = Vec::with_capacity(serialized_size(value, info)?);

    let mut serializer = Serializer::new(&mut vec, info);
    value.serialize(&mut serializer)?;

    Ok(vec)
}

pub fn to_bytes(value: &impl Serialize, info: &'static MessageInfo) -> Result<Bytes, Error> {
    let mut bytes = BytesMut::with_capacity(serialized_size(value, info)?);

    let mut serializer = Serializer::new(&mut bytes, info);
    value.serialize(&mut serializer)?;

    Ok(bytes.freeze())
}

pub fn to_bytes_mut(value: &impl Serialize, info: &'static MessageInfo) -> Result<BytesMut, Error> {
    let mut bytes = BytesMut::with_capacity(serialized_size(value, info)?);

    let mut serializer = Serializer::new(&mut bytes, info);
    value.serialize(&mut serializer)?;

    Ok(bytes)
}

pub fn serialized_size(value: &impl Serialize, info: &'static MessageInfo) -> Result<usize, Error> {
    let mut size_hint = SizeHint::new(info);
    value.serialize(&mut size_hint)
}
