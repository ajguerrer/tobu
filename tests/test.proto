syntax = "proto3";

package tests;

message Outer {
	int32 s1 = 1;
	Inner s2 = 2;
}

message Inner {
	string i2 = 1;
}

