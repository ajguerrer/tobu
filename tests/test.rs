use serde::Serialize;
use tobu::{
    encoding::field::FieldNumber,
    info::{Cardinality, FieldInfo, MessageInfo, Syntax, Type},
};

#[derive(Serialize)]
struct Inner {
    i1: String,
}

#[derive(Serialize)]
struct Outer {
    s1: i32,
    s2: Inner,
}

const OUTER_INFO: MessageInfo = MessageInfo {
    syntax: Syntax::Proto3,
    name: "Outer",
    fields: &[
        FieldInfo {
            name: "s1",
            json_name: "s1",
            type_name: "s1",
            number: unsafe { FieldNumber::new_unchecked(1) },
            cardinality: Cardinality::Optional,
            ty: Type::Int32,
            message_info: None,
            enum_info: None,
            oneof_index: None,
            packed: true,
        },
        FieldInfo {
            name: "s2",
            json_name: "s2",
            type_name: "s2",
            number: unsafe { FieldNumber::new_unchecked(2) },
            cardinality: Cardinality::Optional,
            ty: Type::Message,
            message_info: Some(&INNER_INFO),
            enum_info: None,
            oneof_index: None,
            packed: true,
        },
    ],
    is_map: false,
};

const INNER_INFO: MessageInfo = MessageInfo {
    syntax: Syntax::Proto3,
    name: "Inner",
    fields: &[FieldInfo {
        name: "i2",
        json_name: "i2",
        type_name: "i2",
        number: unsafe { FieldNumber::new_unchecked(1) },
        cardinality: Cardinality::Optional,
        ty: Type::String,
        message_info: None,
        enum_info: None,
        oneof_index: None,
        packed: true,
    }],
    is_map: false,
};

#[test]
fn main() {
    let s = Outer {
        s1: 1,
        s2: Inner {
            i1: "hello".to_string(),
        },
    };
    let size = tobu::serialized_size(&s, &OUTER_INFO).unwrap();
    assert_eq!(size, 10);
}
