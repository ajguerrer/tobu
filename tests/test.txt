file {
  name: "test.proto"
  package: "tests"
  message_type {
    name: "Outer"
    field {
      name: "s1"
      number: 1
      label: LABEL_OPTIONAL
      type: TYPE_INT32
      json_name: "s1"
    }
    field {
      name: "s2"
      number: 2
      label: LABEL_OPTIONAL
      type: TYPE_MESSAGE
      type_name: ".tests.Inner"
      json_name: "s2"
    }
  }
  message_type {
    name: "Inner"
    field {
      name: "i2"
      number: 1
      label: LABEL_OPTIONAL
      type: TYPE_STRING
      json_name: "i2"
    }
  }
  syntax: "proto3"
}
