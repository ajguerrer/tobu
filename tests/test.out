
k

test.prototests"5
Outer
s1 (Rs1
s2 (2.tests.InnerRs2"
Inner
i2 (	Ri2bproto3